### PR TITLE
Raising exceptions in test mode (Issue: #10079) 

### PR DESCRIPTION
**Pull request by SahilC**

### CODE REVIEW

File: src/atom-environment.coffee [hunk 1]
```diff
@@ -171,6 +171,7 @@ class AtomEnvironment extends Model
       devMode, @configDirPath, resourcePath, safeMode, @config, styleManager: @styles,
       commandRegistry: @commands, keymapManager: @keymaps, notificationManager: @notifications,
       grammarRegistry: @grammars, deserializerManager: @deserializers, viewRegistry: @views
+      testMode: @inTestMode()
```
Comment (50Wliu) [inline src/atom-environment.coffee]: To be consistent with the rest I would change this to be `@testMode` instead.

**Comment by SahilC** [comment]
I've made the requested changes, could someone verify if everything is ok?

**Comment by thomasjo** [comment]
Can you explain the purpose of this PR? How is this different to [spec mode](https://github.com/atom/atom/blob/8229924b96497f0d0d0db47355d3e821c58a77ec/src/atom-environment.coffee#L427-L429)?

**Comment by thomasjo** [comment]
OK, I see from the title that this is meant to fix #10079. Next time, please add all relevant details to the description body of the PR. I'll add some review comments for the remainder of my review.

File: src/atom-environment.coffee [hunk 2]
```diff
@@ -171,6 +171,7 @@ class AtomEnvironment extends Model
       devMode, @configDirPath, resourcePath, safeMode, @config, styleManager: @styles,
       commandRegistry: @commands, keymapManager: @keymaps, notificationManager: @notifications,
       grammarRegistry: @grammars, deserializerManager: @deserializers, viewRegistry: @views
+      testMode: @testMode
```
Comment (thomasjo) [inline src/atom-environment.coffee]: This needs to be removed. We already have this in the form of `specMode`.

File: src/atom-environment.coffee [hunk 3]
```diff
@@ -420,6 +421,10 @@ class AtomEnvironment extends Model
   inDevMode: ->
     @devMode ?= @getLoadSettings().devMode
 
+  # Public: Returns a {Boolean} that is `true` if the current window is in test mode.
+  inTestMode: ->
+    @testMode ?= @getLoadSettings().testMode
+
```
Comment (thomasjo) [inline src/atom-environment.coffee]: Same as earlier, needs to be removed.

File: src/package-manager.coffee [hunk 4]
```diff
@@ -30,7 +30,7 @@ module.exports =
 class PackageManager
   constructor: (params) ->
     {
-      configDirPath, @devMode, safeMode, @resourcePath, @config, @styleManager,
+      configDirPath, @devMode, safeMode, @testMode, @resourcePath, @config, @styleManager,
```
Comment (thomasjo) [inline src/package-manager.coffee]: `@testMode` needs to be removed.

File: src/package-manager.coffee [hunk 5]
```diff
@@ -397,7 +397,7 @@ class PackageManager
 
       options = {
         path: packagePath, metadata, packageManager: this, @config, @styleManager,
-        @commandRegistry, @keymapManager, @devMode, @notificationManager,
+        @commandRegistry, @keymapManager, @devMode, @testMode, @notificationManager,
```
Comment (thomasjo) [inline src/package-manager.coffee]: Bye bye, `@testMode`

File: src/package.coffee [hunk 6]
```diff
@@ -32,7 +32,7 @@ class Package
   constructor: (params) ->
     {
       @path, @metadata, @packageManager, @config, @styleManager, @commandRegistry,
-      @keymapManager, @devMode, @notificationManager, @grammarRegistry, @themeManager,
+      @keymapManager, @devMode, @testMode, @notificationManager, @grammarRegistry, @themeManager,
```
Comment (thomasjo) [inline src/package.coffee]: You know the drill by now 😄

File: src/package.coffee [hunk 7]
```diff
@@ -124,7 +124,10 @@ class Package
             else
               @activateNow()
           catch error
-            @handleError("Failed to activate the #{@name} package", error)
+            if @testMode
```
Comment (thomasjo) [inline src/package.coffee]: Change `@testMode` to `@specMode`.

**Review by thomasjo — CHANGES_REQUESTED** [review]

**Comment by 50Wliu** [comment]
Ah hah, I thought there was a specMode already defined somewhere...but I didn't see anywhere in the diff :P.

**Comment by SahilC** [comment]
Yeah, I was confused about that too! Should I just rename everything to specMode?

**Comment by thomasjo** [comment]
@SahilC It already exists, so remove almost all additions (as per my review comments). Keep the change package.coffee, but rename the `@testMode` there to `@specMode`.

**Comment by SahilC** [comment]
But specMode isn't being passed to package.coffee's constructor. Am I missing something?

**Comment by thomasjo** [comment]
@SahilC Sorry I should have looked at the diffs more closely, but you can use [`atom.inSpecMode()`](https://atom.io/docs/api/v1.11.2/AtomEnvironment#instance-inSpecMode).

File: src/package.coffee (hunk 7 above)
Comment (thomasjo) [inline src/package.coffee]: Sorry I was mistaken earlier, but `@testMode` can be replaced with `atom.inSpecMode()`.

**Comment by SahilC** [comment]
I've removed all occurrences of testMode, and replaced the check with atom.inSpecMode(). Wouldn't it be more consistent to use a variable rather than actually calling that function directly?

**Comment by thomasjo** [comment]
I don't think it's worth adding the necessary changes for that given that we only check if we're in spec mode on a single line. I'd argue we should rather remove `@devMode` when we get around to converting this to JS.

This looks good to me now!

**Review by thomasjo — APPROVED** [review]

**Comment by SahilC** [comment]
Cool, thanks!

File: src/package.coffee [hunk 8]
```diff
@@ -124,7 +124,10 @@ class Package
             else
               @activateNow()
           catch error
-            @handleError("Failed to activate the #{@name} package", error)
+            if atom.inSpecMode()
+              raise error
```
Comment (maxbrunsfeld) [inline src/package.coffee]: This should be `throw error`.

**Comment by maxbrunsfeld** [comment]
Thanks for looking into this problem @SahilC! Just one comment 👆 .

**Comment by joh1234** [comment]
Great, thanks. https://goo.gl/JBljdR

**Pull request closed without merging**